### ISSUE TITLE
UnsupportedOperationException w/ the Gradient Example

### ISSUE THREAD

**Issue by conner985** [Protelis/Protelis#118]
Hello,

I receive this error when I try  to run a simulation with the attached files:

`16:16:46.532 [Thread-3] ERROR i.u.a.core.implementations.Engine - The simulation engine crashed.
java.lang.UnsupportedOperationException: null
	at java.util.Collections$UnmodifiableMap.put(Unknown Source)
	at org.protelis.vm.util.StackImpl.put(StackImpl.java:61)
	at org.protelis.vm.impl.AbstractExecutionContext.putVariable(AbstractExecutionContext.java:130)
	at org.protelis.lang.interpreter.impl.RepCall.eval(RepCall.java:66)
	at org.protelis.lang.interpreter.impl.All.eval(All.java:59)
	at org.protelis.vm.impl.SimpleProgramImpl.compute(SimpleProgramImpl.java:81)
	at org.protelis.vm.ProtelisVM.runCycle(ProtelisVM.java:45)
	at it.unibo.alchemist.model.implementations.actions.RunProtelisProgram.execute(RunProtelisProgram.java:140)
	at it.unibo.alchemist.model.implementations.reactions.AReaction.execute(AReaction.java:285)
	at it.unibo.alchemist.core.implementations.Engine.doStep(Engine.java:164)
	at it.unibo.alchemist.core.implementations.Engine.run(Engine.java:368)
	at java.lang.Thread.run(Unknown Source)`

Let me know if you also need the Java source that played the simulation.

[resources.zip](https://github.com/Protelis/Protelis/files/890354/resources.zip)

**Comment by DanySK** [comment]
This error generates from Alchemist, but is clearly Protelis failing. The root cause seems to me to be the `All` not correctly creating a stack frame. This IMHO is only a problem for computations that start contain a single `rep` in the main script.

@jakebeal, I'd solve this by creating a root stack frame upon program initialization in `AbstractExecutionContext`. The alternative is to force the stack frame creation even for single-lined `All` bodies, which sounds like an overkill to me.

I am also trying to reproduce the bug in the regression test suite

**Comment by DanySK** [comment]
I reproduced the issue in a local, small test case:
```protelis
// EXPECTED_RESULT: $CYCLE
rep(val <- 0) {
	val + 1
}
```

**Comment by DanySK** [comment]
@jakebeal, I solved this with a pointwise change. Basically, we were pushing a stack frame id onto the stack, but we were not preparing it for being writable. Since it is very small, if you have no objections I'd like to push it directly to develop, along with the related regression test -- here is the diff

```diff
diff --git a/protelis/protelis-interpreter/src/main/java/org/protelis/vm/impl/AbstractExecutionContext.java b/protelis/protelis-interpreter/src/main/java/org/protelis/vm/impl/AbstractExecutionContext.java
index 0f65d51..ea60608 100644
--- a/protelis/protelis-interpreter/src/main/java/org/protelis/vm/impl/AbstractExecutionContext.java
+++ b/protelis/protelis-interpreter/src/main/java/org/protelis/vm/impl/AbstractExecutionContext.java
@@ -104,11 +104,11 @@ public abstract class AbstractExecutionContext implements ExecutionContext {
         }
         assert previousRoundTime != null : "Round time is null.";
         callStack.clear();
-        callStack.add((byte) 1);
         env.setup();
         toSend = newLinkedHashMapWithExpectedSize(exportsSize);
         gamma = new StackImpl(functions);
         theta = Collections.unmodifiableMap(nm.getNeighborState());
+        newCallStackFrame((byte) 1);
     }
 
     @Override
```

**Comment by jakebeal** [comment]
I think I understand what's happened, and I'm OK with you pushing that change. Can you please also add a test rep03 that validates the correct behavior with your new test case?

**Comment by DanySK** [comment]
Fixed and regression tested in 0720990e1619c8792db4a0eb9d59f196b3ec79f4.
